luwak/src/main/java/uk/co/flax/luwak/BatchMonitor.java

```diff
@@ -0,0 +1,194 @@
+
+/**
+ * A Monitor contains a set of MonitorQuery objects, and runs them against
+ * passed-in InputDocuments.
+ */
+public class BatchMonitor {
+
+    private Monitor monitor;
+    private long commitTimeout;
+    private long lastCommitTime = 0;
+    private volatile ScheduledFuture<?> commitJob = null;
+    private final ScheduledExecutorService commitExecutor;
+
+    private Map<String, Indexable> pendingUpdates = new ConcurrentHashMap<>();
+    private Queue<String> pendingDeletes = new ConcurrentLinkedQueue<>();
+
+    /**
+     * Create a new BatchMonitor instance, that writes to the supplied monitor
+     *
+     * Note that when the BatchMonitor is closed, the embedded Monitor
+     * (and its all components) will also be closed.
+     *
+     * @param monitor the underlying monitor to write to
+     * @param configuration the MonitorConfiguration
+     * @throws IOException on IO errors
+     */
+    public BatchMonitor(Monitor monitor, BatchQueryIndexConfiguration configuration) throws IOException {
+        this.monitor = monitor;
+        this.commitTimeout = configuration.getCommitTimeout();
+        this.commitExecutor = Executors.newSingleThreadScheduledExecutor();
+    }
+
+    /**
+     * Set the commit timeout in milliseconds.
+     *
+     * @param timeout time in milliseconds after which a record will be committed.
+     */
+    public void setCommitTimeout(long timeout) {
+        this.commitTimeout = timeout;
+    }
+
+    private void scheduleCommit() {
+        // If we are not bunching up commits, then this is a no-op.
+        if (commitTimeout == 0) {
+            return;
+        }
+
+        // If we have a pending commit job, then we are done,
+        // that pending commit will pick up the changes
+        if (commitJob != null && !commitJob.isDone()) {
+            return;
+        }
+
+        // It maybe that lastCommitTime + commitTimeout - System.currentTimeMillis()
+        // is negative, which is fine, the job will just run ASAP then.
+        if (commitTimeout > 0) {
+            commitJob = this.commitExecutor.schedule(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        {
+                            String[] deletes = pendingDeletes.toArray(new String[0]);
+                            if (pendingDeletes.removeAll(Arrays.asList(deletes))) {
+                                monitor.deleteById(deletes);
+                            }
+                        }
+                        {
+                            String[] keys = pendingUpdates.keySet().toArray(new String[0]);
+                            Indexable[] updates = pendingUpdates.values().toArray(new Indexable[0]);
+                            if (pendingUpdates.keySet().removeAll(Arrays.asList(keys))) {
+                                monitor.commit(Arrays.asList(updates));
+                            }
+                        }
```

Comment:
Hm, what happens here if the pending updates add a query and then delete it?  This is going to run things in the wrong order, isn't it?  It would be good to get some more comprehensive tests that mix up adds and deletes.